### PR TITLE
SEO: deindex /en, fix redirects, canonicals and soft-404s (GSC audit)

### PR DESCRIPTION
**Pull request by christosporios**
## Context

A Google Search Console audit (2026-07-15) of `sc-domain:opencouncil.gr` found **12.8K not-indexed vs 2.17K indexed pages**, and — the headline problem — **Greek searchers being served `/en` pages**: 8.16K impressions on `/en/` URLs in 3 months, 4,812 of them from users in Greece (119 clicks).

Root causes found in the code, each fixed in its own commit:

### 1. Deindex `/en` (`fix(seo): canonicalize all locale variants...` + sitemap commit)
Every `/en` page self-canonicalized and the sitemap advertised every `/en` URL as an hreflang alternate, so Google indexed the English tree and served it to Greek users while the Greek variants sat in the 6.5K "Discovered – currently not indexed" backlog. `buildHreflangAlternates(path, locale)` is now `buildCanonicalAlternates(path)`: every locale variant canonicalizes to the realm's default-locale (unprefixed) URL, no hreflang cluster is emitted, and the sitemap no longer lists `/en` alternates. `/en` stays fully usable for humans.

### 2. `www` → apex 301 (`feat(proxy)`)
`www.opencouncil.gr` served the whole site as a duplicate host. The proxy now 301s exact `www.<realm-domain>` hosts to the apex, preserving path and query.

### 3. Foreign locale prefix → 301 (`feat(proxy)`)
`/fr/...` rendered fine on the `.gr` host (and `/el/...` on `.fr`), creating an orphaned duplicate tree (e.g. `/fr/docs` in GSC). On known realm hosts, the other realm's default-locale prefix now 301s to the unprefixed URL. `/en` keeps serving 200; localhost keeps all locales for development.

### 4. Phantom `/meetings/` redirects (`fix(seo)`)
Until e686908e (2026-05-29) the sitemap emitted `/{cityId}/meetings/{id}` URLs — the real routes have no `/meetings/` segment — leaving ~1.9K GSC 404s that Google keeps recrawling. Those (and locale-prefixed variants) now 301 to the real URLs. A negative lookahead protects the real `/api/meetings/*` routes.

### 5. Soft-404s for dead meetings/subjects (`fix(seo): stop soft-404s...`)
Dead subject IDs (subjects are regenerated when a meeting is reprocessed) returned HTTP 200 + streamed 404 UI + JS-injected noindex → 518 "excluded by noindex" + soft-404s in GSC.
- **Dead meetings** now throw `notFound()` from the meeting layout's `generateMetadata` (above the `loading.tsx` Suspense boundary) → **real HTTP 404 for everyone** (verified with curl).
- **Dead subjects** render below that boundary, so a real 404 is impossible without dropping the loading state; they now return explicit `robots: noindex, nofollow` metadata, and `Googlebot` was added to `htmlLimitedBots` so crawlers get blocking (non-streamed) metadata — the noindex and every canonical tag land in the `<head>` instead of being streamed late in the body.

### 6. Missing canonicals (`feat(seo): add canonical metadata...`)
~1.2K pages were "duplicate without user-selected canonical" because they had no metadata at all: privacy, terms, corrections, docs, explain (its client component was extracted so the page can be a server component), the fullscreen map layout, the city parties tab and the city notification-signup page now emit canonicals.

### 7. noindex private pages (`feat(seo)`)
`/unsubscribe` (token link), `/notifications/[id]` (personalized) and `/offer-letter/[offerId]` (private) are now `noindex, nofollow`.

## Verification

All verified against a local dev server with spoofed `Host` headers:
- `/en/about` canonical → `https://opencouncil.gr/about`; zero `hreflang` links; sitemap has zero `xhtml:link` entries
- `/athens/meetings/abc[/subjects/def]` → 301 to real URL; `/api/meetings/upcoming` untouched (200)
- `Host: www.opencouncil.gr` `/athens?x=1` → 301 `https://opencouncil.gr/athens?x=1`
- `Host: opencouncil.gr` `/fr/athens` → 301 `/athens`; `/en/athens` stays 200; plain localhost `/fr/...` stays 200
- Dead meeting → HTTP 404 (all UAs); dead subject → `noindex, nofollow` in the blocking head for Googlebot UA and in the streamed head for browsers; real meeting/subject pages unaffected
- `npx tsc --noEmit` clean, `npx jest` 977 passed (21 new tests for the canonical helper and redirect logic), eslint clean on changed files

Note: `www` redirect in production also depends on DNS/TLS for the `www` subdomain resolving to the app (verified reachable today).

## After merging

- GSC will take weeks to re-crawl; the "Duplicate/noindex/404" report counts should start dropping, and `/en` impressions for searchers in Greece should trend to zero.
- Consider resubmitting the sitemap in GSC to accelerate re-crawling.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Every request passes through new proxy redirect branches and site-wide canonical/hreflang behavior changes affect all indexed URLs; logic is well-tested and mostly metadata/redirect only, with no auth or data-model changes.
> 
> **Overview**
> Addresses a Google Search Console audit where Greek users were served `/en` URLs and thousands of pages sat in duplicate, 404, or noindex states.
> 
> **Canonical and sitemap:** `buildHreflangAlternates` is replaced by `buildCanonicalAlternates` across public pages so every locale variant points at the realm’s unprefixed default URL; hreflang clusters are removed and the sitemap lists only those default-locale URLs.
> 
> **Redirects:** `next.config.mjs` adds 301s from legacy `/{city}/meetings/{id}` paths to real meeting URLs. `proxy.ts` 301s exact `www.<realm>` hosts to the apex and strips foreign default-locale prefixes on the wrong host (`/fr` on `.gr`, `/el` on `.fr`).
> 
> **Crawlers and dead content:** `htmlLimitedBots` is extended with `Googlebot` so meeting `notFound()` in layout `generateMetadata` yields a real HTTP 404 for bots; dead subjects return blocking `noindex` metadata and clear inherited OG/canonical. Private or token pages (`unsubscribe`, notification view, offer letter) are `noindex`. Several static and city tab routes gain missing canonical metadata.
> 
> **Tests:** New unit tests cover canonical URL building and SEO redirect helpers.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 92a29b725f56b4014ecf4395b716b3cd59f69ff8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<h3>Greptile Summary</h3>

This PR resolves a Google Search Console audit showing 12.8K non-indexed pages and English (`/en`) URLs being served to Greek users. It makes site-wide SEO corrections across seven distinct areas: canonical consolidation, proxy-level redirects, dead-content 404s, noindex for private pages, sitemap cleanup, `alternateLinks` suppression, and missing canonical metadata.

- **Canonical/hreflang:** `buildHreflangAlternates` is replaced by `buildCanonicalAlternates` across all public pages; every locale variant now points its canonical at the realm's unprefixed default URL, hreflang clusters are removed, and both the sitemap and next-intl's `alternateLinks` header are suppressed.
- **Proxy redirects:** `proxy.ts` gains www→apex and foreign-locale-strip 301s; `next.config.mjs` adds 301s for ~1.9K phantom `/meetings/` URLs.
- **Dead content:** `notFound()` in the meeting layout's `generateMetadata` produces real HTTP 404s for crawlers; dead subjects return explicit `noindex` with inherited canonical/OG tags nulled out.

<h3>Confidence Score: 5/5</h3>

Safe to merge — changes are metadata, redirect, and canonical-only with no auth or data-model mutations.

Every redirect path has dedicated tests pinned to the actual config and module logic. The notFound() in generateMetadata follows Next.js documented behavior verified with curl. The hreflang-to-canonical refactor is complete across all 43 files with no missed call sites. No data, auth, or API surface changes.

No files require special attention. The most complex file is src/proxy.ts where redirect ordering matters; the ordering is correct and commented.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/proxy.ts | Adds www→apex 301 and foreign-locale-prefix 301 as early-exit branches in the correct order before the .fr rewrite and i18n middleware. |
| src/lib/seo-redirects.ts | New pure edge-safe module with wwwRedirectTarget and foreignLocaleRedirectPath; fully covered by the new test file. |
| src/lib/utils/hreflang.ts | buildHreflangAlternates replaced by buildCanonicalAlternates; returns only { canonical } pointing at the realm's unprefixed base URL. |
| next.config.mjs | Adds phantom /meetings/ → real URL 301 redirects with a negative lookahead protecting /api/meetings/*; htmlLimitedBots documented against next@16.2.6. |
| src/app/sitemap.ts | Removes buildAlternates helper and all alternates fields; sitemap now lists only default-locale unprefixed URLs. |
| src/app/[locale]/(city)/[cityId]/(meetings)/[meetingId]/layout.tsx | notFound() thrown from generateMetadata so htmlLimitedBots receive a real HTTP 404 for dead meeting URLs. |
| src/app/[locale]/(city)/[cityId]/(meetings)/[meetingId]/subjects/[subjectId]/page.tsx | Dead subjects return robots:noindex with alternates/openGraph/twitter nulled; page body independently calls notFound() for the UI. |
| src/i18n/routing.ts | alternateLinks: false suppresses next-intl's automatic hreflang Link response header. |
| src/lib/realm.ts | Adds foreignLocalesForRealm helper; correctly excludes 'en' since it is shared across realms. |
| src/lib/__tests__/seo-redirects.test.ts | Comprehensive coverage of both redirect helpers including edge cases for trailing slash, port, case, bare locale, and partial segment matching. |
| src/lib/__tests__/next-config-redirects.test.ts | Pins next.config.mjs redirect patterns using Next's own path-to-regexp; guards the /api/meetings/* negative lookahead. |

<sub>Reviews (6): Last reviewed commit: ["feat(seo): cover the pages that had neit..."](https://github.com/schemalabz/opencouncil/commit/a8487fda54670c0f4670834e3caadb454a175f81) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44509479)</sub>

### CODE REVIEW

**Comment by christosporios** [comment]
I [force-pushed](https://github.com/schemalabz/opencouncil/compare/9bd00a26ff6884fa32e510545c05a3cd54073843..35ecfad91d8d649d7f8932d63a9b4aed6607502b) to rebase onto main and address review comments:

- Rebased onto main. The `/explain` page was rewritten on main (#539) as a server component with its own `generateMetadata`, which made this PR's client-component extraction obsolete — dropped `explain-content.tsx` and instead adapted the new page to the renamed `buildCanonicalAlternates` helper (this also resolves the trailing-newline nit, since the file no longer exists).
- `wwwRedirectTarget` now strips trailing slashes, so `www.../athens/` reaches the apex in one 301 instead of chaining into the trailing-slash 308 (per greptile). Test added.
- Documented that the `htmlLimitedBots` list is copied from next@16.2.6 and needs a resync on Next upgrades (per greptile).
- No change for the Bugbot finding on the `/api/meetings/*` redirect — it doesn't reproduce; see reply on the thread.

Re-verified after the rebase: `tsc` clean, 978 tests pass, and curl checks against a local server confirm all redirect/canonical/noindex behaviors, including `/api/meetings/upcoming` → 200 (not redirected) and single-hop `www` redirects.

File: next.config.mjs [hunk 1]
```diff
@@ -64,6 +64,32 @@ const nextConfig = {
...
+            // filesystem) and defers prefixed variants to the locale rules.
+            {
+                source: '/:cityId((?!api/|en/|el/|fr/)[^/]+)/meetings/:meetingId/subjects/:subjectId',
+                destination: '/:cityId/:meetingId/subjects/:subjectId',
+                statusCode: 301,
+            },
+            {
+                source: '/:cityId((?!api/|en/|el/|fr/)[^/]+)/meetings/:meetingId',
+                destination: '/:cityId/:meetingId',
+                statusCode: 301,
```
[the remark this answers is not in the document]
Comment (christosporios) [reply next.config.mjs:95]: Doesn't reproduce — the negative lookahead evaluates against the remaining path string at the parameter's start position, not the isolated segment, so for `/api/meetings/upcoming` the lookahead sees `api/meetings/upcoming`, `(?!api/)` fails, and the rule doesn't match. Verified against a running server: `curl -s -o /dev/null -w '%{http_code} redirect=%{redirect_url}' localhost:3000/api/meetings/upcoming` → `200 redirect=` (route handler executes, no redirect), while `/athens/meetings/abc` → `301 /athens/abc` and `/apitest/meetings/x` (a segment merely *starting* with `api`) still redirects as intended.

File: src/proxy.ts [hunk 2]
```diff
@@ -31,6 +32,15 @@ export default async function proxy(req: NextRequest) {
...
     }
 
+    // www hosts duplicate the apex domain in Google's index — 301 them away.
+    const wwwTarget = wwwRedirectTarget(
+        req.headers.get('host'),
+        req.nextUrl.pathname + req.nextUrl.search,
+    );
+    if (wwwTarget) {
+        return NextResponse.redirect(wwwTarget, 301);
+    }
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/proxy.ts:43]: Fixed — `wwwRedirectTarget` now takes `pathname`/`search` separately and strips trailing slashes (root `/` preserved), so `www.../athens/?x=1` → `https://opencouncil.gr/athens?x=1` in a single 301. Covered by a new test and verified with curl.

File: next.config.mjs [hunk 3]
```diff
@@ -83,6 +109,14 @@ const nextConfig = {
         ];
     },
     skipTrailingSlashRedirect: true,
+    // Serve blocking (non-streamed) metadata to these crawlers, so notFound()
+    // thrown in generateMetadata produces a real HTTP 404 instead of a streamed
+    // 200 + noindex (GSC was reporting dead subject/meeting URLs as soft-404s).
+    // Overriding replaces Next's default HTML_LIMITED_BOT_UA_RE, so this is the
+    // default list copied wholesale with `Googlebot` prepended — plain Googlebot
+    // is deliberately absent from Next's default because it executes JS, but a
+    // streamed 404 still reaches it as HTTP 200.
```
[the remark this answers is not in the document]
Comment (christosporios) [reply next.config.mjs:132]: Documented — the comment now notes the list is copied from next@16.2.6 (`next/dist/shared/lib/router/utils/html-bots.js`) and should be resynced on Next upgrades. Went with the documented copy over importing the private dist path: a deep import would auto-track upstream but couples the config to Next internals that can move without notice, turning a routine upgrade into a config-load failure.

File: src/app/[locale]/(other)/explain/explain-content.tsx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                    >
+                        <Link href="/about">
+                            Αν είστε στην αυτοδιοίκηση, πατήστε εδώ
+                        </Link>
+                    </Button>
+                </motion.section>
+            </div>
+        </div>
+    );
+}
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/app/[locale]/(other)/explain/explain-content.tsx]: Moot after the rebase — main's #539 rewrote /explain as a server component with its own generateMetadata, so the extraction (and this file) was dropped; the page now just uses the renamed canonical helper.

File: src/app/[locale]/(city)/[cityId]/(other)/notifications/page.tsx [hunk 5]
```diff
@@ -4,11 +4,16 @@ import { OnboardingStage } from '@/lib/types/onboarding';
...
-    title: "Εγγραφή για ενημερώσεις | OpenCouncil",
-    description: "Εγγραφείτε για να λαμβάνετε ενημερώσεις για θέματα που συζητούνται στα δημοτικά συμβούλια",
-};
+export async function generateMetadata(props: { params: Promise<{ cityId: string }> }): Promise<Metadata> {
+    const params = await props.params;
+    return {
+        title: "Εγγραφή για ενημερώσεις | OpenCouncil",
+        description: "Εγγραφείτε για να λαμβάνετε ενημερώσεις για θέματα που συζητούνται στα δημοτικά συμβούλια",
+        alternates: await buildCanonicalAlternates(`/${params.cityId}/notifications`),
+    };
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/app/[locale]/(city)/[cityId]/(other)/notifications/page.tsx:22]: Fixed — `generateMetadata` now fetches the city via `getCityCached` and throws `notFound()` when it's missing, same as the parties page. Verified: `/nocity/notifications` → 404 with no canonical tag; a real city still gets 200 + canonical. (The `supportsNotifications` redirect case needs no metadata guard: the body's `redirect()` produces a 307, so the metadata never renders.)

File: src/app/[locale]/(city)/[cityId]/(meetings)/[meetingId]/subjects/[subjectId]/page.tsx [hunk 6]
```diff
@@ -15,7 +15,17 @@ export async function generateMetadata(
...
+        // reprocessed) can't produce a real 404: the page renders below the
+        // meeting loading.tsx Suspense boundary, so the 200 shell is already
+        // flushed when notFound() throws. Explicit noindex metadata is the
+        // reliable signal — htmlLimitedBots (next.config.mjs) puts it in the
+        // blocking <head> for crawlers. The page body still calls notFound()
+        // for the UI.
+        return {
+            title: 'Not Found',
+            robots: { index: false, follow: false },
+        };
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/app/[locale]/(city)/[cityId]/(meetings)/[meetingId]/subjects/[subjectId]/page.tsx:34]: Fixed — the dead-subject metadata now explicitly nulls `alternates`, `openGraph` and `twitter` so the meeting layout's tags no longer merge in. Verified with curl (Googlebot UA): the dead-subject response has `noindex, nofollow`, zero canonical tags, and its og:title is a generic "Not Found" instead of the parent meeting's — so there's no canonical for the noindex signal to leak onto. Real subject pages unchanged.

**Comment by christosporios** [comment]
I [force-pushed](https://github.com/schemalabz/opencouncil/compare/ef14025a8bc4e8bbe790633f6afa11f91ab4243a..92a29b725f56b4014ecf4395b716b3cd59f69ff8) to address Bugbot's follow-up: dead-subject metadata now explicitly clears the inherited `alternates`/`openGraph`/`twitter` from the meeting layout, so a noindex subject URL no longer carries the parent meeting's canonical and social tags. Folded into the soft-404 commit; tsc clean, all 978 tests pass, verified with curl.

**Comment by kouloumos** [comment]
I [force-pushed](https://github.com/schemalabz/opencouncil/compare/92a29b725f56b4014ecf4395b716b3cd59f69ff8..547794e8eb4f5e7d6e261068280da5ab5496d6c4) to rebase on main now that #542 landed:

- `next.config.mjs`: kept both redirect groups — the `/map` redirects from main plus this PR's phantom-`/meetings/` rules
- dropped the `(fullscreen)/map/layout.tsx` hunk from the canonical-metadata commit — the file no longer exists and canonicalizing a deleted page is moot; the commit's other pages are untouched
- no other changes; `tsc` and the seo-redirects/hreflang suites pass on the rebased branch

Original branch state preserved at [`backup/pr-541-original`](https://github.com/kouloumos/opencouncil/tree/backup/pr-541-original).

**Review by kouloumos — APPROVED** [review]
This is careful work, and the verification section made reviewing it much easier — I re-ran the checks independently and everything holds, including against the pr-541 preview: `/fr/athens` 301s to `/athens` (so the foreign-locale strip provably covers preview hosts too), the phantom `/athens/meetings/…` URLs 301 to the real ones, the sitemap emits zero hreflang entries, and `/api/meetings/upcoming` stays 200 — which settles the Bugbot lookahead dispute in your favour with a live check on top of a regex-level one (compiling the redirect sources with Next's bundled path-to-regexp shows the lookahead evaluates against the raw remaining path, so `/api/…` can never match; there's also no city or meeting id in the DB that could collide with the `/meetings/` patterns).

One thing was undermining the whole plan though: next-intl's middleware emits the full hreflang cluster as an HTTP `Link` response header on every page — on by default, and header hreflang counts for Google the same as tags or sitemap entries. So while the HTML now says "the canonical is the unprefixed URL", every response header still said "`/en/x` is a distinct English alternate — index it", and on `.gr` it kept advertising `/fr/*` URLs that this PR 301s away. Easy to miss since it never appears in the HTML. Verified on the preview: both signals present side by side on `/athens` and `/en/athens`. Fixed with 016705c9 (`alternateLinks: false` in the routing config) — verified on a dev server that the header is gone while the canonicals and the locale-strip redirects are unchanged.

I pushed that alongside three more commits that extend the PR in its own direction — each independently droppable if you disagree (`git revert --no-edit a8487fda c9c31431 c746c702 016705c9`):

- a8487fda covers the pages that had neither a canonical nor a noindex, so their `/en`/`/fr` variants carried no signal at all: canonicals for the two petition pages, noindex for `/profile`, `/sign-in`, the `/admin` tree (once, in its layout), the embed pages (via the embed layout), the city widget page, the presentation view, and the meeting admin/settings pages — the last two also null the canonical they inherit from the meeting layout. Verified on a dev server.
- c9c31431 pins the redirect edges in tests: the disputed `/api/meetings/*` lookahead (patterns extracted from `next.config.mjs` itself and compiled with Next's own path-to-regexp, so this never gets re-litigated), mixed-case `www` hosts, and `/elefsina` on the `.fr` host not being mistaken for an `/el` prefix.
- c746c702 drops the params plumbing the `buildCanonicalAlternates` rename left unused.

The preview is rebuilding with these commits now — once it's up I'll re-run the full matrix against it (the hreflang header should be gone from `/` and `/en/*` while everything else stays identical, plus the new canonical/noindex coverage) and then we merge.

**Comment by kouloumos** [comment]
Re-ran the matrix against the rebuilt preview: the hreflang `Link` header is gone from `/athens` and `/en/athens`, while everything else behaves exactly as before — canonicals, the locale-strip and phantom-`/meetings/` redirects, the sitemap, and `/api/meetings/upcoming`. The new coverage from a8487fda is also live (`/petition` canonical, `noindex` on `/sign-in` and `/admin`).

All green — merging.
